File: examples/client-gen/tictactoe/instructions/play.py
from __future__ import annotations
import typing
from solders.pubkey import Pubkey
from solders.instruction import Instruction, AccountMeta
import borsh_construct as borsh
from .. import types
from ..program_id import PROGRAM_ID


class PlayArgs(typing.TypedDict):
    tile: types.tile.Tile


layout = borsh.CStruct("tile" / types.tile.Tile.layout)


class PlayAccounts(typing.TypedDict):
    game: Pubkey
    player: Pubkey


def play(
    args: PlayArgs,
    accounts: PlayAccounts,
    program_id: Pubkey = PROGRAM_ID,
    remaining_accounts: typing.Optional[typing.List[AccountMeta]] = None,
) -> Instruction:
    keys: list[AccountMeta] = [
        AccountMeta(pubkey=accounts["game"], is_signer=False, is_writable=True),
        AccountMeta(pubkey=accounts["player"], is_signer=True, is_writable=False),
    ]
    if remaining_accounts is not None:
        keys += remaining_accounts
    identifier = b"\xd5\x9d\xc1\x8e\xe48\xf8\x96"
    encoded_args = layout.build(
        {
            "tile": args["tile"].to_encodable(),
        }
    )
    data = identifier + encoded_args
    return Instruction(program_id, data, keys)

File: ts-reference/examples/tic-tac-toe/generated-client/instructions/play.ts
import { TransactionInstruction, PublicKey } from "@solana/web3.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import BN from "bn.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@project-serum/borsh" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as types from "../types" // eslint-disable-line @typescript-eslint/no-unused-vars
import { PROGRAM_ID } from "../programId"

export interface PlayArgs {
  tile: types.TileFields
}

export interface PlayAccounts {
  game: PublicKey
  player: PublicKey
}

export const layout = borsh.struct([types.Tile.layout("tile")])

export function play(args: PlayArgs, accounts: PlayAccounts) {
  const keys = [
    { pubkey: accounts.game, isSigner: false, isWritable: true },
    { pubkey: accounts.player, isSigner: true, isWritable: false },
  ]
  const identifier = Buffer.from([213, 157, 193, 142, 228, 56, 248, 150])
  const buffer = Buffer.alloc(1000)
  const len = layout.encode(
    {
      tile: types.Tile.toEncodable(args.tile),
    },
    buffer
  )
  const data = Buffer.concat([identifier, buffer]).slice(0, 8 + len)
  const ix = new TransactionInstruction({ keys, programId: PROGRAM_ID, data })
  return ix
}
